lib/Sema/SemaType.cpp

```diff
@@ -3465,49 +3465,59 @@ static void checkNullabilityConsistency(TypeProcessingState &state,
-                                              constArrTy->getSizeModifier(),
+    ASTContext &Context = S.Context;
+    SplitQualType split = T.split();
+    const Type *ty = split.Ty;
+    const ArrayType *arrTy = cast<ArrayType>(ty);
+    QualType elemTy = makeNestedArrayChecked(S, arrTy->getElementType(), Loc);
+    switch (ty->getTypeClass()) {
+      case Type::ConstantArray: {
+        const ConstantArrayType *constArrTy = cast<ConstantArrayType>(ty);
+        return Context.getConstantArrayType(elemTy,
+                                            constArrTy->getSize(),
+                                            constArrTy->getSizeModifier(),
+                                            T.getCVRQualifiers(),
+                                            true);
+      }
+      case Type::IncompleteArray: {
+        const IncompleteArrayType *incArrTy = cast<IncompleteArrayType>(ty);
+        return Context.getIncompleteArrayType(elemTy,
+                                              incArrTy->getSizeModifier(),
                                               T.getCVRQualifiers(),
                                               true);
-        }
-        case Type::IncompleteArray: {
-          const IncompleteArrayType *incArrTy = cast<IncompleteArrayType>(ty);
-          return Context.getIncompleteArrayType(elemTy,
-                                                incArrTy->getSizeModifier(),
-                                                T.getCVRQualifiers(),
-                                                true);
-        }
-        case Type::DependentSizedArray:
-        case Type::VariableArray: 
-          // checked versions of these arrays cannot be created. An error
-          // message is produced by BuildArrayType for the outer error type
-          // because these result in the outer array type being
-          // dependently-typed or variably-sized.
-          break;
-        default:
-           assert("unexpected array type");
-           break;
       }
+      case Type::DependentSizedArray:
+      case Type::VariableArray: 
+        // Checked versions of these arrays cannot be created. An error
+        // message is produced by BuildArrayType for the outer error type
+        // because these result in the outer array type being
+        // dependently-typed or variably-sized.
+        break;
+      default:
+          assert("unexpected array type");
+          break;
     }
+  } else if (const TypedefType *TD = dyn_cast<TypedefType>(T)) {
+    const ArrayType *AT = dyn_cast<ArrayType>(T.getCanonicalType());
+    if (AT != nullptr && !AT->isChecked())
+      S.Diag(Loc, diag::err_checked_array_of_unchecked_array) << TD->getDecl();
+  } else if (const ParenType *PT = dyn_cast<ParenType>(T)) {
+    assert(!T.hasLocalQualifiers());
+    QualType innerType = makeNestedArrayChecked(S, PT->getInnerType(), Loc);
+    ASTContext &Context = S.Context;
+    return Context.getParenType(innerType);
```

Comment:
Naive question: is there any more generic way of figuring out if the type is a trivial wrapper for another one and grabbing the inner type?  If ever a new type "wrapper" were to be created, we'd miss that case here, right?